src/common/libevent.hpp

```diff
@@ -463,6 +466,150 @@ class Bufferevent : public NonCopyable, public NonMovable {
+                             int flags = DNS_QUERY_NO_SEARCH) {
+        // callback viene tenuta viva in quanto viene copiata nello scope
+        auto cb = new std::function<void(int, char, int, int, void *)>(
+            [callback](int r, char t, int c, int ttl, void *addresses) {
+                callback(r, t, c, ttl, ipv4_address_list(c, addresses));
+            });
+        if (resolve(base->dns_base, name.c_str(), flags, handle_resolve, cb) ==
+            nullptr) {
+            MK_THROW(EvdnsBaseResolveIpv4ExceptionError);
+        }
+    }
+
+    static std::vector<std::string> ipv6_address_list(int count,
+                                                      void *addresses) {
+        std::vector<std::string> results;
+        static const int size = 16;
+        if (count >= 0 && count <= INT_MAX / size + 1) {
+            char string[128]; // Is wide enough (max. IPv6 length is 45 chars)
+            for (int i = 0; i < count; ++i) {
+                // Note: address already in network byte order
+                if (inet_ntop(AF_INET6, (char *)addresses + i * size, string,
+                              sizeof(string)) == nullptr) {
+                    break;
+                }
+                results.push_back(string);
+            }
+        }
+        return results;
+    }
+
+    template <
+        decltype(evdns_base_resolve_ipv6) resolve = ::evdns_base_resolve_ipv6>
+    static void resolve_ipv6(Var<EvdnsBase> base, std::string name,
+                             ResolveCallback callback,
+                             int flags = DNS_QUERY_NO_SEARCH) {
+        auto cb = new std::function<void(int, char, int, int, void *)>(
+            [callback](int r, char t, int c, int ttl, void *addresses) {
+                callback(r, t, c, ttl, ipv6_address_list(c, addresses));
+            });
+        if (resolve(base->dns_base, name.c_str(), flags, handle_resolve, cb) ==
+            nullptr) {
+            MK_THROW(EvdnsBaseResolveIpv6ExceptionError);
+        }
+    }
+
+    static std::vector<std::string> ptr_address_list(int count,
+                                                     void *addresses) {
+        std::vector<std::string> results;
+        results.push_back(std::string(*(char **)addresses));
+        return results;
+    }
+
+    static in_addr *ipv4_pton(std::string address, in_addr *netaddr) {
+        if (inet_pton(AF_INET, address.c_str(), netaddr) != 1) {
+            throw InvalidIPv4AddressError();
+        }
+        return (netaddr);
+    }
+
+    static in6_addr *ipv6_pton(std::string address, in6_addr *netaddr) {
+        if (inet_pton(AF_INET6, address.c_str(), netaddr) != 1) {
+            throw InvalidIPv6AddressError();
+        }
+        return (netaddr);
+    }
+
+    static void resolve_reverse(Var<EvdnsBase> base, std::string address,
+                                ResolveCallback callback,
+                                int flags = DNS_QUERY_NO_SEARCH) {
```

Comment:
Speaking about design, I'm now wondering whether it was a good idea to use static methods. Given that here we do not pass `base` to the lambda to keep it alive, what do we gain by using a static method?

Assuming that we do not want to pass `base` to the lambda, it would then suffice to have a constructor that returns a `Var<EvdnsBase>` and, from then one, if one can call a method of `Var<EvdnsBase>` it is because it holds a reference to the shared pointer, hence the object is alive. Hence the constructor is critical. But, if we want to force the programmer to always use a smart pointer, we're probably better off with all static methods, because this removes any possibility of using the class in another way. This makes me wonder whether a better style would be, instead, to write ordinary functions named just like libevent functions, except that they take C++11 arguments and live inside the `mk` namespace. After all, if we're set to write wrappers, ordinary functions with some C++11 goodies is the least amount of change with respect to what we are wrapping, that is, libevent.

Speaking of whether to pass `base` to the lambda, I don't have a good answer for now.